openssl-dynamic/src/main/c/sslcontext.c

```diff
@@ -1659,6 +1538,153 @@ TCN_IMPLEMENT_CALL(void, SSLContext, setCertVerifyCallback)(TCN_STDARGS, jlong c
     }
 }
 
+/**
+ * Returns an array containing all the X500 principal's bytes.
+ *
+ * Partly based on code from conscrypt:
+ * https://android.googlesource.com/platform/external/conscrypt/+/master/src/main/native/org_conscrypt_NativeCrypto.cpp
+ */
+static jobjectArray principalBytes(JNIEnv* e, const STACK_OF(X509_NAME)* names) {
+    jobjectArray array;
+    jbyteArray bArray;
+    int count;
+    int length;
+    unsigned char *buf;
+    X509_NAME* principal;
+
+    if (names == NULL) {
+        return NULL;
+    }
+
+    count = sk_X509_NAME_num(names);
+    if (count <= 0) {
+        return NULL;
+    }
+
+    array = (*e)->NewObjectArray(e, count, byteArrayClass, NULL);
+    if (array == NULL) {
+        return NULL;
+    }
+
+    for (int i = 0; i < count; i++) {
+        principal = sk_X509_NAME_value(names, i);
+        buf = NULL;
+        length = i2d_X509_NAME(principal, &buf);
+        if (length < 0) {
+            // We need to delete the local references so we not leak memory as this method is called via callback.
+            OPENSSL_free(buf);
+
+            // In case of error just return an empty byte[][]
+            return (*e)->NewObjectArray(e, 0, byteArrayClass, NULL);
+        }
+        bArray = (*e)->NewByteArray(e, length);
+        (*e)->SetByteArrayRegion(e, bArray, 0, length, (jbyte*) buf);
+        (*e)->SetObjectArrayElement(e, array, i, bArray);
+
+        // Delete the local reference as we not know how long the chain is and local references are otherwise
+        // only freed once jni method returns.
+        (*e)->DeleteLocalRef(e, bArray);
+        OPENSSL_free(buf);
+    }
+
+    return array;
+}
+
+/**
+ * Partly based on code from conscrypt:
+ * https://android.googlesource.com/platform/external/conscrypt/+/master/src/main/native/org_conscrypt_NativeCrypto.cpp
+ */
+static int cert_requested(SSL* ssl, X509** x509Out, EVP_PKEY** pkeyOut) {
+    tcn_ssl_ctxt_t *c = SSL_get_app_data2(ssl);
+    char ssl2_ctype = SSL3_CT_RSA_SIGN;
+    int ctype_num;
+    jbyte* ctype_bytes;
+    jobjectArray issuers;
+    JNIEnv *e;
+    jbyteArray keyTypes;
+    X509* certificate;
+    EVP_PKEY* privatekey;
+    tcn_get_java_env(&e);
+
+    /* Clear output of key and certificate in case of early exit due to error. */
+    *x509Out = NULL;
+    *pkeyOut = NULL;
+
+#if !defined(OPENSSL_IS_BORINGSSL) && OPENSSL_VERSION_NUMBER < 0x1000200fL
+    switch (ssl->version) {
+        case SSL2_VERSION:
+            ctype_bytes = (jbyte*) &ssl2_ctype;
+            ctype_num = 1;
+            break;
+        case SSL3_VERSION:
+        case TLS1_VERSION:
+        case TLS1_1_VERSION:
+        case TLS1_2_VERSION:
+        case DTLS1_VERSION:
+            ctype_bytes = (jbyte*) ssl->s3->tmp.ctype;
+            ctype_num = ssl->s3->tmp.ctype_num;
+            break;
+    }
+#else
+    const uint8_t* ctype = NULL;
```

Comment:
openssl will cast this to a `char*` under the hood anyways ... can we just pass in `ctype_bytes` directly?